src/transformers/generation/utils.py

```diff
@@ -3365,34 +3354,80 @@ def _beam_search(
             `return_dict_in_generate=True` or a [`~generation.GenerateBeamEncoderDecoderOutput`] if
             `model.config.is_encoder_decoder=True`.
         """
-        # init values
+
+        def flatten_beam_dim(tensor):
+            """3D with beams in a separate dim -> 2D with batch*beam in batch dim"""
+            shape = list(tensor.shape)
+            return torch.reshape(tensor, [shape[0] * shape[1]] + shape[2:])
+
+        def unflatten_beam_dim(tensor, batch_size, num_beams):
+            """2D with batch*beam in batch dim -> 3D with beams in a separate dim"""
+            shape = list(tensor.shape)
+            return torch.reshape(tensor, [batch_size, num_beams] + shape[1:])
+
+        def gather_beams(nested: Union[torch.Tensor, Iterable], beam_indices: torch.Tensor):
+            """
+            Gathers the beam slices indexed by beam_indices into new beam array.
+
+            Args:
+                nested (`Iterable` or `torch.Tensor`): A tensor or an iterable of tensors containing data to be
+                    gathered. The tensor is a 2D or a 3D tensor with the two first dimensions depicting the batch and
+                    the beam dimensions.
+                beam_indices (`torch.Tensor` of shape `(batch_size, num_beams_to_select)`): The indices of the beams to
+                    select .
+
+            Returns:
+                A tensor or an iterable of tensors with the selected beams
+            """
+
+            def gather_fn(tensor, beam_indices):
+                # `take_along_dim` requires its indices arg to have the same number of dims as `input`
+                while len(beam_indices.shape) < len(tensor.shape):
+                    beam_indices = beam_indices.unsqueeze(-1)
+                gathered_tensor = torch.take_along_dim(input=tensor, indices=beam_indices, dim=1)
```

Comment:
```suggestion
                num_unsqueezes = tensor.dim() - beam_indices.dim()
                beam_indices = beam_indices.view(*beam_indices.shape, *([1] * num_unsqueezes))
                gathered_tensor = torch.take_along_dim(input=tensor, indices=beam_indices, dim=1)
```

Reply:
cf: https://x.com/jxmnop/status/1885378534102671520

Reply:
this is not a bad pattern, see Horace He's reply: https://x.com/cHHillee/status/1885422843384783320 :p 

IMO, the original version is more readable, I'd rather leave as is